### ISSUE TITLE
Improved notification display

### ISSUE THREAD

**Issue by kevincoleman** [online-go/online-go.com#879]
The display of notifications could be optimized for a better user experience. This, of course, is up for discussion first and foremost. Things I’ve noticed that could improve:
- notifications display in multiple places
- notifications appear an a location that’s fairly UI-blocking, depending on how many there are
- notifications are not formatting in a very mobile friendly way
- notifications have linked text as the main action (could be improved with clear buttons, icons, &c.

I’ll be revisiting this issue with screenshots shortly.

I’d be happy to spearhead this effort with UI mock-ups aiming at solving these problems, if you’re open to it.

Edit:
Here are some next steps I identified further down in the conversation here. I could use some help on some of them, as I’m not as familiar with the inner workings of the app yet. If nobody disagrees, I think we could treat them like action items:

- [ ] catalog the types of notifications we have/want [Spreadsheet](https://docs.google.com/spreadsheets/d/1Sd4-U6D2dq4ay50B4xBe7yPDgiGHeEuoNWSZagMGWLY/edit?usp=sharing)
- [ ] catalog users’ concerns [Spreadsheet](https://docs.google.com/spreadsheets/d/1F60PjfxqRheI4NeNT1k5pzfQHbAmZazDzfG5srq3yQA/edit?usp=sharing), [Forum thread](https://forums.online-go.com/t/too-many-pop-ups-in-the-initial-page/23620)
- [ ] create a working draft of an ideal user experience that serves both needs well and is extensible
- [ ] follow up here to talk implementation details before attempting to write it.

**Comment by anoek** [comment]
:+1:  I'm definitely open to exploring some UI/UX improvements for notifications.

**Comment by kevincoleman** [comment]
![Image](https://i.imgur.com/ygHDQNo.jpg)

Notifications stack up on mobile.

Probably should just pick one way to display.

**Comment by kevincoleman** [comment]
So I haven’t done a lot of research yet, but it seems to me that we might get a good response to adding an icon with a number badge near the top right, Facebook-style. My reasoning is that many notifications are not immediately actionable for a normal user just logging on to make moves.

Toaster style notifications are strong because they’re highly visible and accessible, but weak because they’re UI-blocking and don’t stack well.

Menu-style notifications are strong because they’re non-UI-blocking and they stack well, but they’re not as visible at first.

I suggest we tackle this by thinking about what types of notifications go best in each place.

The main type of notifications I see are the “announcement” style. They’re something users likely didn’t ask for but which benefit from high visibility. These properties actually make them feel a lot like ads. It would probably be best to manage these kinds of notifications in a way that puts the user first, so they don’t feel spammed. Some ideas might be: move these notifications to the aforementioned menu location; allow users to customize what notification types they see; limit the amount of time these kinds of notifications show.

Other types of helpful notifications might be: it’s your move, friend requests, game start/end/invite, chat-related, and others, I expect. Most of these seem like they could also easily go in the menu.

We may still get some value from toaster style notifications for things that are immediately urgent. Things like “You have only 1 minute left to play in X game” (when that game is not the game showing on screen. I would expect all toaster notifications to go away in under 60 or 30 seconds.

The main benefits:
- Increase use trust by helping them feel less spammed
- Stop blocking the UI whenever possible
- Help users triage notifications more efficiently to complete all that are actionable more completely
- others?

Potential drawbacks:
- decreased exposure for current toaster notifications because they’re no longer as visible
- users could turn off notifications that we want them to see in order to encourage community activity
- others?

**Comment by kevincoleman** [comment]
And by “add a menu with a badge” I mean “better utilize the menu & badge”

This also sets us up for using that badge number on a PWA or native app icon.

**Comment by BHydden** [comment]
This discussion has been reignited in [the forums](https://forums.online-go.com/t/too-many-pop-ups-in-the-initial-page/23620?u=bhydden).

**Comment by kevincoleman** [comment]
That seems good—this needs discussion before we implement. As with any feature set, though, it needs a certain level of executive decision when we move forward. So far I’ve seen great management of this app’s feature set, so I have high confidence.

It seems like what needs to happen is:
- catalog the types of notifications we have/want
- catalog users’ concerns
- create a working draft of an ideal user experience that serves both needs well and is extensible
- follow up here to talk implementation details before attempting to write it.

I’ll try my best to start organizing this data so we can move forward responsibly.

**Comment by kevincoleman** [comment]
![Screen Shot 2019-12-20 at 11 37 20 PM](https://user-images.githubusercontent.com/445707/71304930-df294a80-2381-11ea-81fd-9c6df78655a2.png)

What’s going on here? This looks like a notification, but can’t even be dismissed?

**Comment by BHydden** [comment]
There were complaints that when the letterbox style notifications were dismissed from the top-right, there was no way to recover their information if it was desired later. 
For the duration of a notification's life, it is also permanently at the top of the "Games" and "Home" pages so that after the pop-up notification has been dismissed, the link can still be accessed.

**Comment by BHydden** [comment]
It's a temporary workaround for the inadequacies of our current system, and probably doesn't need to be included in the upgrades we're discussing here, so long as their function is retained in a more ideal implementation.

**Comment by kevincoleman** [comment]
Hm. This feels like a step backward. I get that in the case where people care about those notifications they could want better persistence, but removing the ability to dismiss them at all is a poor solution as it creates a more frustrating UX for people who don’t want them. It moves these notifications firmly from the “sort of spammy notification” class to the “I paid for no advertising what is this about” class.

While I can see that it might not need a more permanent fix this second, I definitely recommend we implement something more elegant here.

**Comment by kevincoleman** [comment]
If nobody here has any objections, I’d like to donate a little time toward creating some UI mockups of how I envision a more elegant solution working. I’ll document my process, and of course it will be up for discussion before we collectively make any big decisions.

Sound good?

**Comment by BHydden** [comment]
Please do ❤️

**Comment by anoek** [comment]
Awesome, looking forward to them @kevincoleman!

**Comment by yewang** [comment]
I suggest getting rid of the "pop-up" style notifications altogether.

They are redundant with the banner at the top of the home screen.

Alternatively, put these notifications along with the other notifications (e.g., for friend requests, game challenges, etc.) that are already hidden inside the drop-down menu.

**Comment by kevincoleman** [comment]
I just updated the issue description with a checklist of acceptance criteria. I also started spreadsheets for the first two items, as I think we need to consider all the cases methodically. Before I dive into trying to mock up a solution I thought it would be prudent to ask for your eyes on those sheets. I want the method to feel very data-informed, so we know why we decide what we decide in each case.

On those spreadsheets, feel free to use the comment feature to strike up discussion on each particular item. This is a pretty big issue, so discussing all its details in a single thread (like here) would get messy fast.

**Comment by BHydden** [comment]
Some new features / functionality have been added regarding notification blocking, but a total notification upgrade is probably still worth looking into 🙂

https://github.com/online-go/online-go.com/pull/1423

**Comment by BHydden** [comment]
A new notification system will be needed at some point, and repeating all this discussion at that time will be redundant. Better to keep this open until that time I think.

**Comment by kevincoleman** [comment]
This issue needs to stay open, even if it just sits more.

**Comment by anoek** [comment]
https://forums.online-go.com/t/too-many-pop-ups-in-the-initial-page/23620/137 for some conversation on the topic

**Comment by kevincoleman** [comment]
I think after a bit more discussion on the forum we have enough clarification to make more directed acceptance criteria for this issue!

https://forums.online-go.com/t/too-many-pop-ups-in-the-initial-page/23620/140?u=blindgorgon

Happy to iterate, but I’m going to update the issue description with these steps. Also, I’m happy to try to take a stab at this code, again knowing I’ll probably need lots of feedback and guidance.

**Comment by shinuito** [comment]
In terms of the figma mockups, I don't take issue with most of it.

However, the notification settings grid of 17x4 options seems like massive overkill. I would wager that many people will interact with very few of these options. Unless something really annoyed me, I don't think I would ever go near that page with that many toggles.

**Comment by kevincoleman** [comment]
Yeah, the switches are a bit overwhelming. I wonder if we could get a better feel by going to checkboxes. The utility of the array of switches is undeniable, but I hear you on how visually daunting it is. I reached for toggle switches as they feel more mobile-friendly (bigger target for fat thumbs).

Anyone have a better idea that still gives the same level of customization?

**Comment by anoek** [comment]
One potential idea would be grouping. For instance, we could have "Game notifications" which covers game starts, ends, and challenges, and "Tournament notifications" which covers invitations, start/finish notification, that sort of thing. If we still want granularity those groupings could be expandable so then you can toggle on/off/mixed the group overall, or you can expand it and toggle the individual settings for that fine grained control if you want.

The other thing that comes to mind is separation of some of these, I don't think we'll ever want game starts / ends notifications in the banner for instance. Or, if we do, maybe we need to revisit how those are displayed in that top banner, I don't think they're very fitting currently but perhaps there's a variation where it would make sense.

**Comment by kevincoleman** [comment]
I agree with both of those ideas. :)

I can try to figure out all those tweaks to this UI, but know that I’m just guessing.
